### DISCUSSION TITLE
Optimistic Concurrency Control or simply wrong

### DISCUSSION THREAD

**Discussion by antonkarsten** [prisma/prisma#10250] [Q&A] [answered]
I am new to Prisma and am looking at Optimistic Concurrency Control. the docs at https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide are quite clear. But it only works when we already have a row in our table that we can version. 

I have a schema with a table/model `Reservable` that holds some kind of object that you can reserve e.g. a hotel room. Furthermoe I have a table/model for a `Reservation` which holds a reference to the reservable and the start and end date.

The problem is that i cannot put a `version` field on `Reservable`, because as I understand the prisma client does not support a nested create in an `updateMany`. (See also #7257 that was discussed between @doddi321 and @ryands17)

here is my schema, I will put my attempted solution below

```prisma
model Order {
  id           Int           @id @default(autoincrement())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  owner        User          @relation(fields: [ownerId], references: [id])
  ownerId      Int
  reservations Reservation[]
}

model Reservation {
  id                Int        @id @default(autoincrement())
  createdAt         DateTime   @default(now())
  updatedAt         DateTime   @updatedAt
  start             DateTime
  end               DateTime
  reservable        Reservable @relation(fields: [reservableId], references: [id])
  reservableId      Int
  order             Order      @relation(fields: [orderId], references: [id])
  orderId           Int
  optimisticCounter Int // this is the field I use for "OCC-versioning"

  @@unique([reservableId, optimisticCounter])
}

// This could be a hotel room, a bicycle etc
model Reservable {
  id           Int           @id @default(autoincrement())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  reservations Reservation[]
}

model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  email     String   @unique
  name      String?
  orders    Order[]
}
```

My solution is as follows:
1. Get the most recent reservation (the one with the highest `optimisticCounter` value)
2. Determine the value of our optimisticCounter for our next reservation
3. Re-check if the reservable is still free (i.e. there are no reservations for this reservable at these dates)
4. Make the actual reservation. This should fail if in the mean time another reservation has happened because of the constraint `@@unique([reservableId, optimisticCounter])`

```typescript
getMostRecentReservationForReservable(reservableId: number) {
  return prisma.reservation.findFirst({
    orderBy: {
      optimisticCounter: "desc",
    },
    where: {
      reservableId: reservableId,
    },
  });
}

async createReservation(
  orderId: number,
  reservable: Reservable,
  start: string,
  end: string
) {
  /**
   * STEP 1: get the most recent reservation (the one with the highest optimisticCounter value)
   */

  const mostRecentReservation =
    await this.getMostRecentReservationForReservable(reservable.id);

  /**
   * STEP 2: Determine the value of our optimisticCounter for our next reservation
   */
  let nextOptimisticCounterValue = 1; // if there are no reservations yet for this reservable, we need to get started somewhere

  // based on the last made reservation, we determine what the id of the next reservation should be
  if (mostRecentReservation) {
    nextOptimisticCounterValue = mostRecentReservation.optimisticCounter + 1;
  }

  /**
   * STEP 3: Re-check if the reservable is still free (i.e. there are no reservations for this reservable at these dates)
   */
  const isReservableFree = await this.isReservableFree(
    reservable.id,
    start,
    end
  );

  if (!isReservableFree) {
    throw Error("This reservable is not available (any more)");
  }

  /**
   * STEP 4: Make the actual reservation
   */
  const reservation = await prisma.reservation.create({
    data: {
      reservableId: reservable.id,
      start,
      end,
      orderId,
      optimisticCounter: nextOptimisticCounterValue,
    },
  });

  return reservation;
}
```

I wrote a number of tests where I injected variable wait times between the steps, and everything seem to work fine. Nevertherless my question remains:
1. Is this at all correct/safe?
2. Is there a simpler solution?
3. Should I better use an interactive transaction?

i am happy with any feedback, thanx!

**Comment by austincrim** [comment, marked as answer]
Hi @antonkarsten 👋, thanks for the question!

Your implementation looks good to us! Like you mentioned, you could also use an [interactive transaction](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) for this, so it is a matter of your preference which one you go with.

You might also want to read through [this issue](https://github.com/prisma/prisma/issues/4988) and add your use case/+1 to help us decide priority. 

Let me know if you have any other questions!

**Reply by antonkarsten** [reply]
@austincrim thank you very much for your answer !

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
